Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeLibs.cs

```diff
@@ -0,0 +1,1669 @@
+    ///      to circumvent this VBE bug.
+    ///
+    /// 2)  The ITypeInfo for base classes of UserForms crash with an AV if you call ITypeInfo::GetDocumentation(MEMBERID_NIL) 
+    ///     to get the type name.  We've got to remember that the VBE didn't ever intend for us to get hold of these objects, 
+    ///     so there will be little bugs.  This bug is also resolved in the provided wrappers.
+    ///
+    /// This class can also be cast to ComTypes.ITypeInfo for raw access to the underlying type information
+    /// </remarks>
+    public class TypeInfoWrapper : ComTypes.ITypeInfo, IDisposable
+    {
+        private DisposableList<TypeInfoWrapper> _typeInfosWrapped;
+        private TypeLibWrapper _containerTypeLib;
+        public TypeLibWrapper Container { get => _containerTypeLib; }
+        private int _containerTypeLibIndex;
+        private bool _isUserFormBaseClass = false;
+        private IntPtr _rawObjectPtr;
+        private ComTypes.ITypeInfo _wrappedObjectRCW;
+
+        private ComTypes.TYPEATTR _cachedAttributes;
+        public ComTypes.TYPEATTR Attributes { get => _cachedAttributes; }
+
+        private RestrictComInterfaceByAggregation<ComTypes.ITypeInfo> _ITypeInfo_Aggregator;
+        private ComTypes.ITypeInfo target_ITypeInfo { get => _ITypeInfo_Aggregator?.WrappedObject ?? _wrappedObjectRCW; }
+
+        private RestrictComInterfaceByAggregation<IVBEComponent> _IVBEComponent_Aggregator;
+        private IVBEComponent target_IVBEComponent { get => _IVBEComponent_Aggregator?.WrappedObject; }
+
+        private RestrictComInterfaceByAggregation<IVBETypeInfo> _IVBETypeInfo_Aggregator;
+        private IVBETypeInfo target_IVBETypeInfo { get => _IVBETypeInfo_Aggregator?.WrappedObject; }
+
+        public bool HasVBEExtensions { get => _IVBETypeInfo_Aggregator?.WrappedObject != null; }
+
+        private bool _hasModuleScopeCompilationErrors;
+        public bool HasModuleScopeCompilationErrors => _hasModuleScopeCompilationErrors;
+
+        /// <summary>
+        /// Exposes an enumerable collection of functions provided by the ITypeInfo
+        /// </summary>
+        public class FuncsCollection : IIndexedCollection<TypeInfoFunc>
+        {
+            TypeInfoWrapper _parent;
+            public FuncsCollection(TypeInfoWrapper parent) => _parent = parent;
+            override public int Count { get => _parent.Attributes.cFuncs; }
+            override public TypeInfoFunc GetItemByIndex(int index) => new TypeInfoFunc(_parent, index);
+
+            public TypeInfoFunc Find(string name, TypeInfoFunc.PROCKIND procKind)
+            {
+                foreach (var func in this)
+                {
+                    if ((func.Name == name) && (func.ProcKind == procKind)) return func;
+                }
+                return null;
+            }
+        }
+        public FuncsCollection Funcs;
+
+        /// <summary>
+        /// Exposes an enumerable collection of variables/fields provided by the ITypeInfo
```

Comment:
A clarification is needed here - all variables/fields or only publicly accessible variables/fields? If all, is there a way to know whether its scope?

Reply:
it exposes whatever the particular ITypeInfo exposes, which at this implementation detail we can't know.